src/DacpacTool/ScriptParser.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.SqlTools.ServiceLayer.BatchParser;
+using Microsoft.SqlServer.Dac.Model;
+
+namespace MSBuild.Sdk.SqlProj.DacpacTool
+{
+    public class ScriptParser : ICommandHandler
+    {
+        private readonly Parser _parser;
+        private bool _parsed;
+        private readonly List<string> _includedFileNames = new List<string>();
+        private readonly StringBuilder _scriptBuilder = new StringBuilder();
+        private const string BATCH_SEPARATOR = "GO";
+
+        public ScriptParser(string sourceFile, IVariableResolver variableResolver)
+        {
+            _parser = new Parser(this, variableResolver, new StreamReader(sourceFile), sourceFile);
+        }
+
+        private void Parse()
+        {
+            if (! _parsed)
+            {
+                _parser.Parse();
+                _parsed = true;
+            }
+        }
+
+        public IEnumerable<string> CollectFileNames()
+        {
+            Parse();
+            return _includedFileNames;
+        }
+
+        public string GenerateScript()
+        {
+            Parse();
+            return _scriptBuilder.ToString();
+        }
+
+        BatchParserAction ICommandHandler.Go(TextBlock batch, int repeatCount, SqlCmdCommand tokenType)
+        {
+            batch.GetText(true, out string batchText, out LineInfo lineInfo);
+            _scriptBuilder.AppendLine(batchText);
+            _scriptBuilder.AppendLine(BATCH_SEPARATOR);
+            return BatchParserAction.Continue;
+        }
+
+        BatchParserAction ICommandHandler.Include(TextBlock includeBlock, out TextReader stream, out string newFilename)
+        {
+            includeBlock.GetText(true, out string includedFileName, out LineInfo lineInfo);
+
+            if (!Path.IsPathRooted(includedFileName))
+            {
+                var position = lineInfo.GetStreamPositionForOffset(0);
+                includedFileName = Path.Combine(Path.GetDirectoryName(position.Filename), includedFileName);
+            }
+
+            _includedFileNames.Add(includedFileName);
+
+            stream = new StreamReader(includedFileName);
+            newFilename = includedFileName;
+            return BatchParserAction.Continue;
+        }
+
+        BatchParserAction ICommandHandler.OnError(Token token, OnErrorAction action)
+        {
+            // Write error to console
+            var error = new ModelValidationError(
+                sourceName: Path.GetFileNameWithoutExtension(token.Filename),
```

Comment:
Why `GetFilenameWithoutExtension()`?

Reply:
I thought about this one for a while, and I could really go either way on it. `sourceName` is meant to be the name of a TSql object, and would typically not include the extension.

Reply:
I went with the full filename in #55 because that gives us traceability to Visual Studio. That way you can double click the error in the Error List and be taken to the file that is causing the error.

Reply:
Okay, if the full filename gives us better functionality, I'll go with that.